smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -58,100 +85,235 @@ class DragonBackend:
+                node_name = hosts[node_num]
+                local_policy = Policy(
+                    placement=Policy.Placement.HOST_NAME, host_name=node_name
+                )
+                tmp_proc = TemplateProcess(
+                    target=request.exe,
+                    args=request.exe_args,
+                    cwd=request.path,
+                    env={**request.current_env, **request.env},
+                    # stdout=Popen.PIPE,
+                    # stderr=Popen.PIPE,
+                    policy=local_policy,
+                )
+                grp.add_process(nproc=request.tasks_per_node, template=tmp_proc)
+
+            grp.init()
+            grp.start()
+            self._group_infos[step_id] = ProcessGroupInfo(
+                process_group=grp,
+                puids=grp.puids,
+                return_codes=[],
+                status=SmartSimStatus.STATUS_RUNNING,
+                hosts=hosts,
+            )
+            self._running_steps.append(step_id)
+            started.append(step_id)
+
+        if started:
+            print(f"{self._updates}: {started=}")
+
+        for step_id in started:
+            self._queued_steps.pop(step_id)
+
+        terminated = []
+        for step_id in self._running_steps:
+            group_info = self._group_infos[step_id]
+            grp = group_info.process_group
+
+            if grp.status == DRG_RUNNING_STATUS:
+                group_info.status = SmartSimStatus.STATUS_RUNNING
             else:
-                updated_statuses[step_id] = (
-                    SmartSimStatus.STATUS_NEVER_STARTED,
-                    return_codes,
+                puids = group_info.puids
+                if puids is not None and all(puid is not None for puid in puids):
+                    try:
+                        group_info.return_codes = [
+                            Process(None, ident=puid).returncode for puid in puids
+                        ]
+                    except (ValueError, TypeError):
+                        group_info.return_codes = [-1 for _ in puids]
+                else:
+                    group_info.return_codes = [0]
+                group_info.status = (
+                    SmartSimStatus.STATUS_FAILED
+                    if any(group_info.return_codes) or grp.status == DRG_ERROR_STATUS
+                    else SmartSimStatus.STATUS_COMPLETED
                 )
 
-        return DragonUpdateStatusResponse(statuses=updated_statuses)
+            if group_info.status in TERMINAL_STATUSES:
+                terminated.append(step_id)
+
+        if terminated:
+            print(f"{self._updates}: {terminated=}", flush=True)
+        for step_id in terminated:
+            self._running_steps.remove(step_id)
+            self._completed_steps.append(step_id)
+            group_info = self._group_infos[step_id]
+            if group_info is not None:
+                for host in group_info.hosts:
+                    with self._hostlist_lock:
```

Comment:
Small "performance" thing: I know we're not really using this lock atm, but I am going to push to reverse the order of these lines, just to save the number of `acquire`s that need to be made.